fix(summary_parser): extract "totals" when "summary" is "0"

hargut opened the pull request:
using client_requests::callgrind::start_instrumentation() and stop_instrumentation() leads to summary with a "0" value skipping summary in that case to use totals which are present

gamma0987 commented:
Due to your newest comments in #242, is this still needed?

hargut commented:
Hi @gamma0987,

currently I'm not using an approach with `client_requests::callgrind::start_instrumentation() or start_instrumentation()`.
I do think it would be helpful to add that PR so anyone trying to work with the `client_requests` and `--instr-atstart=no` could get summary values.

Kind regards,
Harald

gamma0987 commented:
Hi, I need a minimal example and the callgrind output files to be able to reproduce your issue.

hargut commented:
Hi @gamma0987,

sorry that I didn't include the example right from the start, I was focused on other topics at the time creating the PR.

Please find a full example within:
https://github.com/hargut/iai-callgrind/tree/fix/output-parsing-client-requests-example
https://github.com/hargut/iai-callgrind/commit/a604585c49e612e87e02599f2730c50423a68b3f

Thank you very much for your consideration.

Have a nice weekend. :smile: 

Kind regards,
Harald

gamma0987 commented:
No problem. Thanks for the example. After being able to reproduce the issue, I'm sure your solution is basically correct. However, you need to make sure that all values in the summary are 0, not just the first one. It's best to do this right after `costs.add_iter_str(...)` based on the `costs`.

gamma0987 commented:
Also, I need a test. It's best to do this in the `benchmark-tests` package with a bench. Setting up a new benchmark test is not documented currently, but you can have a look at the structure in `benchmark-tests/benches` and copy one of the existing ones. A new helper binary can be added in `benchmark-tests/src/helper`.

gamma0987 commented:
I've quickly written a small introduction for the `benchmark-tests` package in `benchmark-tests/README.md`. I hope this helps a little bit.

hargut commented:
Hi @gamma0987,

just added the benchmark test and updated this branch to contain the details.
Unfortunately the test fails here with some error related to the parsing of the stdout before asserting.

```bash
cargo run --package benchmark-tests --profile=bench --bin bench -- bench test_bin_client_requests
```

```
     Running `/home/hargut/Sources/github.com/iai-callgrind/target/release/deps/test_bin_client_requests-8a6ce8c0d303a0b1 --nocapture --bench`
bench: STDOUT:
test_bin_client_requests::client_requests::run parse_output:() -> target/release/client-requests
Hello World.
- end of stdout/stderr
  Instructions:                 465|N/A             (*********)
  L1 Hits:                      609|N/A             (*********)
  L2 Hits:                        0|N/A             (*********)
  RAM Hits:                      85|N/A             (*********)
  Total read+write:             694|N/A             (*********)
  Estimated Cycles:            3584|N/A             (*********)
thread 'main' panicked at benchmark-tests/src/bench.rs:331:17:
Assertion of stdout failed: Diff < left / right > :
 test_bin_client_requests::client_requests::run parse_output:() -> target/release/client-requests
 Hello World.
 - end of stdout/stderr
<  Instructions:                    |N/A             (*********)
<  L1 Hits:                         |N/A             (*********)
<  L2 Hits:                         |N/A             (*********)
<  RAM Hits:                        |N/A             (*********)
<  Total read+write:                |N/A             (*********)
<  Estimated Cycles:                |N/A             (*********)
>  Instructions:                 465|N/A             (*********)
>  L1 Hits:                      609|N/A             (*********)
>  L2 Hits:                        0|N/A             (*********)
>  RAM Hits:                      85|N/A             (*********)
>  Total read+write:             694|N/A             (*********)
>  Estimated Cycles:            3584|N/A             (*********)
```

For some reason the function `filter_stdout` removes all the values before comparing the results.
https://github.com/iai-callgrind/iai-callgrind/blob/40bcc07f2da0dc588edfaa14959ae3b3956cce24/benchmark-tests/src/bench.rs#L356


I also had a look at the patch and the output file details. Running callgrind with the `--instr-atstart=no` and using `client_requests` leads to `.out` files using `valgrind-3.22.0` which look like:
```
# callgrind format

 <...>

events: Ir Dr Dw I1mr D1mr D1mw ILmr DLmr DLmw
summary: 0
```

The patch supplied only checks the summary value on the stripped part, and checks with `.eq` for ` 0`.
So in total the line is only skipped when it is exactly "summary: 0". With the used valgrind version I've not seen any other output than exactly `summary: 0` when using the scenario. Therefore I'm not sure if it is required to check for all values being `0` after splitting the string.

Did your setup produce a different `summary:` output using the example?

Kind regards,
Harald

gamma0987 commented:
> For some reason the function filter_stdout removes all the values before comparing the results.

Please have a look at the `benchmark-tests/README` on the main branch, the basics for the benchmark tests are explained there. Or, have a look at an `expected_stdout` file from another benchmark test.

Please check all values to be not zero. This covers more cases and I want a consistency check in the case all values are zero. Looks like the devs of valgrind also do it that way.

hargut commented on benchmark-tests/Cargo.toml:
```diff
@@ -21,7 +21,7 @@ cargo_metadata = { workspace = true }
 glob = { workspace = true }
 iai-callgrind = { path = "../iai-callgrind", features = [
-  "client_requests_defs",
+  "client_requests_defs", "client_requests"
```
updated.

hargut commented on benchmark-tests/Cargo.toml:
```diff
@@ -195,3 +199,8 @@ path = "benches/test_lib_bench/file_parameter/test_lib_bench_file_parameter.rs"
+
+[[bench]]
+harness = false
+name = "test_bin_client_requests"
```
renamed.

hargut commented:
> > For some reason the function filter_stdout removes all the values before comparing the results.
> 
> Please have a look at the `benchmark-tests/README` on the main branch, the basics for the benchmark tests are explained there. Or, have a look at an `expected_stdout` file from another benchmark test.

Adjusted accordingly. At the moment this test is therefore more of a documentation, as the test case does not validate the actual change. The test will also succeed with a version that does not have the change present.
 
> Please check all values to be not zero. This covers more cases and I want a consistency check in the case all values are zero. Looks like the devs of valgrind also do it that way.

Adjusted, and added a trace output.

hargut commented on iai-callgrind-runner/src/runner/callgrind/summary_parser.rs:
```diff
@@ -30,6 +30,10 @@ impl Parser for SummaryParser {
             if let Some(stripped) = line.strip_prefix("summary:") {
                 trace!("Found line with summary: '{}'", line);
                 costs.add_iter_str(stripped.split_ascii_whitespace());
+                if costs.iter().any(|(_c, u)| *u == 0) {
```
Sorry, that was the intention but apparently selected the wrong method...
Fixed.

gamma0987 commented:
Please use `nightly` rustfmt.

hargut commented:
Updated the PR with all the formatting checked as within `just check-fmt-all`.
